src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java

```diff
@@ -0,0 +1,232 @@
+   * @throws InterruptedException any error in case of retry suspension state
+   */
+  public SuccessFactorsResponseContainer callSuccessFactorsWithRetry(URL endpoint)
+    throws TransportException, InterruptedException, IOException {
+
+    LOG.debug(ResourceConstants.DEBUG_CALL_SERVICE_START.getMsgForKey(SuccessFactorsService.DATA));
+    Response res = retrySapTransportCall(endpoint, MediaType.APPLICATION_JSON);
+    LOG.debug(ResourceConstants.DEBUG_CALL_SERVICE_END.getMsgForKey(SuccessFactorsService.DATA));
+
+    try {
+      return prepareResponseContainer(res);
+    } catch (IOException ioe) {
+      res.close();
+      throw new TransportException(ResourceConstants.ERR_CALL_SERVICE_FAILURE.getMsgForKey(), ioe);
+    }
+  }
+
+  /**
+   * Calls the given URL with retry logic.
+   *
+   * @param endpoint  record fetch URL
+   * @param mediaType mediaType for Accept header property
+   * @return {@code Response}
+   * @throws IOException if all retries fail
+   */
+  public Response retrySapTransportCall(URL endpoint, String mediaType) throws IOException {
+    final Response[] res = new Response[1];
+    Callable<Boolean> fetchRecords = () -> {
+      res[0] = transport(endpoint, mediaType);
+      if (res != null && res[0] != null && res[0].code() >= HttpURLConnection.HTTP_INTERNAL_ERROR) {
+        throw new RetryableException();
+      }
+      return true;
+    };
+
+    Retryer<Boolean> retryer = RetryerBuilder.<Boolean>newBuilder()
+      .retryIfExceptionOfType(RetryableException.class)
+      .withWaitStrategy(WaitStrategies.exponentialWait(WAIT_TIME, TimeUnit.SECONDS))
+      .withStopStrategy(StopStrategies.stopAfterAttempt((int) MAX_NUMBER_OF_RETRY_ATTEMPTS))
+      .build();
+
+    try {
+      retryer.call(fetchRecords);
+    } catch (RetryException | ExecutionException e) {
+      LOG.error("Data Recovery failed for URL {}.", endpoint);
+      throw new IOException();
+    }
+
+    return res[0];
+  }
+
+  /**
+   * Make an HTTP/S call to the given URL.
+   *
+   * @param endpoint  SuccessFactors URL
+   * @param mediaType mediaType for Accept header property
+   * @return {@code Response}
+   * @throws IOException        any http client exceptions
+   * @throws TransportException any error while preparing the {@code OkHttpClient}
+   */
+  private Response transport(URL endpoint, String mediaType) throws IOException, TransportException {
+    OkHttpClient enhancedOkHttpClient = getConfiguredClient().build();
+    Request req = buildRequest(endpoint, mediaType);
+
+    return enhancedOkHttpClient.newCall(req).execute();
+  }
+
+  /**
+   * Prepares the {@code SuccessFactorsResponseContainer} from the given {@code Response}.
+   *
+   * @param res {@code Response}
+   * @return {@code SuccessFactorsResponseContainer}
+   * @throws IOException any IO exception while setting up the response body bytes
+   */
+  private SuccessFactorsResponseContainer prepareResponseContainer(Response res) throws IOException {
```

Comment:
why does this throw an exception?

Reply:
While preparing the response stream in `prepareResponseContainer()`
`res.body().bytes()` throws IOException